### ISSUE TITLE
Issue with response context

### ISSUE THREAD

**Issue by stompweb** [WP-API/node-wpapi#146]
Sorry, I'm not sure if the title of this is correct but it was as a result with a Twitter conversation with @kadamwhite 

https://twitter.com/stompweb/status/699669003669803008

The code that is breaking is located inside my project - specifically here:

https://github.com/stompweb/wp-react-js/blob/master/app/components/Main.js#L37

I have checked the network tab in inspector and I am getting posts back, however when I console.log it then I get an error:

_TypeError: Cannot set property '_paging' of null(…)_

**Comment by kadamwhite** [comment]
My assumption is that the way we are binding the functions that interact with the provided callbacks is affecting the internal context within the request handler methods; I am working on a unit test to demonstrate whether that's the actual bug.

**Comment by kadamwhite** [comment]
Hmm. My assumption appears to be invalid; [this works in my integration test suite](https://github.com/kadamwhite/wordpress-rest-api/pull/148/files#diff-988d0724270de39211e5f42f347ca280R383) (linked code copied below):

```js
	it( 'can GET posts with a context-bound callback', function( done ) {
		function Ctor() {}
		Ctor.prototype.setState = function( state ) {
			this.state = state;
		};
		Ctor.prototype.request = function( cb ) {
			var self = this;
			wp.posts().get(function( err, data ) {
				expect( err ).to.be.null;

				// Context is maintained
				expect( this ).to.equal( self );
				this.setState({
					data: data
				});

				expect( this ).to.have.property( 'state' );
				expect( this.state ).to.be.an( 'object' );
				expect( this.state ).to.have.property( 'data' );
				expect( this.state.data ).to.be.an( 'array' );
				expect( this.state.data.length ).to.equal( 10 );
				cb();
			}.bind( this ) );
		};
		( new Ctor() ).request( done );
	});
```

@stompweb could you paste the entire text of the error you're getting in the console, or screenshot the devtools error report?

**Comment by nabeards** [comment]
Possibly related: I'm getting an empty 'body' in my responses from the Version 2.0-beta12 of the WP plugin. Instead, the expected body appears to be in the 'text' key.

E.g.
wp.posts()

(output at wp-request.js line 124)
req.text = "[actual array of the first page of posts]"
req.body = {}

My full response is an object with no posts and just the _paging key:
 { _paging: 
   { total: '159',
     totalPages: '16',
     links: { next: '{url}/wp-json/wp/v2/posts?page=2' },
     next: 
      { _options: [Object],
        _params: {},
        _supportedMethods: [Object],
        _path: {},
        _template: '' } } }

**Comment by nabeards** [comment]
Well, I just disabled another plugin, and now it works fine. Apparently that plugin is removing the response from 'body' and only leaving it in 'text'.

**Comment by kadamwhite** [comment]
Looking at your response it sounds like the bug isn't part of this library; I will close this. If you discover anything further please feel free to re-open

**Comment by SotirisValogiannis** [comment]
I have the same issue. After some time debugging it I found out that superagent on lib/client.js in definition of `Response.prototype._parseBody` checks if the `Content-Type` of response is json (`this.type`). Wordpress, from what I see in my case, returns `text/html; charset="UTF-8"` something that superagent doesn't like and return `null`.

**Comment by kadamwhite** [comment]
Hmm, @SotirisValogiannis Can you clarify what you're querying within WordPress that is returning text/html?

**Comment by SotirisValogiannis** [comment]
I tried to build make a use of the first example of node-wpapi

```
wp.posts().get(function( err, data ) {
    if ( err ) {
        // handle err
    }
    // do something with the returned posts
});
```

and what I get in console is:

```
err TypeError: Cannot set property '_paging' of null
    at paginateResponse (eval at <anonymous> (bundle.js?ver=0.1:2012), <anonymous>:133:22)
    at returnBody (eval at <anonymous> (bundle.js?ver=0.1:2012), <anonymous>:232:9)
    at tryCatch (eval at <anonymous> (bundle.js?ver=0.1:977), <anonymous>:412:12)
    at invokeCallback (eval at <anonymous> (bundle.js?ver=0.1:977), <anonymous>:427:13)
    at publish (eval at <anonymous> (bundle.js?ver=0.1:977), <anonymous>:395:7)
    at MutationObserver.flush (eval at <anonymous> (bundle.js?ver=0.1:977), <anonymous>:121:5)
```

**Comment by kadamwhite** [comment]
@SotirisValogiannis If you do

```
console.log( wp.posts().toString() );
```

can you open that URL in your browser and see data? What looks like it's happening is that `result.body` is null, which would mean that superagent can't parse the resource; that likely means that WordPress is returning the wrong data, or that

**Comment by SotirisValogiannis** [comment]
Yes. Also using the example code above the server returns the data correctly. I attach you a screenshot: 
![screen shot 2017-02-10 at 03 01 16 2](https://cloud.githubusercontent.com/assets/3330708/22809807/74d59a92-ef3d-11e6-8542-05df23397635.png)

**Comment by kadamwhite** [comment]
@SotirisValogiannis Thanks for your quick response. Could I ask you to paste in the headers for that response? Will likely have a follow-up Q after

**Comment by SotirisValogiannis** [comment]
@kadamwhite sure.

```
Access-Control-Allow-Headers:Authorization, Content-Type
Access-Control-Expose-Headers:X-WP-Total, X-WP-TotalPages
Allow:GET
Cache-Control:no-cache, must-revalidate, max-age=0
Connection:Keep-Alive
Content-Length:1724
Content-Type:text/html; charset="UTF-8"
Date:Fri, 10 Feb 2017 01:12:20 GMT
Expires:Wed, 11 Jan 1984 05:00:00 GMT
Keep-Alive:timeout=5, max=98
Last-Modified:
Link:<http://localhost/wordpress/wp-json/>; rel="https://api.w.org/"
Server:Apache/2.4.23 (Unix) OpenSSL/1.0.2j PHP/5.6.28 mod_perl/2.0.8-dev Perl/v5.16.3
Vary:Accept-Encoding,Cookie
X-Content-Type-Options:nosniff
X-Powered-By:PHP/5.6.28
X-Robots-Tag:noindex
X-WP-Total:1
X-WP-TotalPages:1
```

**Comment by kadamwhite** [comment]
If I may ask, what other plugins do you have running; and what server are you using? I'm trying to replicate the text/html response to test a possible workaround.

**Comment by kadamwhite** [comment]
Also, how are you loading the plugin: is it browserified/webpacked, or are you just enqueuing the wpapi.js library?

**Comment by SotirisValogiannis** [comment]
Your comment helped since the problem was the plugin WP Super Cache, the only which was running. I deactivate it and now works fine :) The problem now is, could I use a cache plugin and node-wpapi at the same time?

**Comment by kadamwhite** [comment]
You should be able to, for sure! Thanks for the tip, I will try to replicate and see if there's something we can do on our end to make that work.

**Comment by SotirisValogiannis** [comment]
As workaround, in case that a json is returned with `text/html`, I used a custom superagent parser like the following:

```
agent.parse['text/html'] = function(text) {
      var obj =  JSON.parse(text);

      return obj instanceof Object ? obj : text
    }
```

**Comment by kadamwhite** [comment]
@SotirisValogiannis I'm trying to replicate the scenario, and installing the caching plugin isn't doing the trick; what version of that plugin are you using? I won't ask you to copy out all of the settings from it, I don't think it's worth it; but I'd be curious to know if `wp.posts()...` works for you if you paste this code into an mu-plugin (e.g. `wp-content/mu-plugins/test.php`):

```
<?php
function wpapi_integration_force_http_content_type( $result, $server, $request ) {
	$result->header( 'Content-Type', 'text/html; charset=' . get_option( 'blog_charset' ) );
	return $result;
}
add_filter( 'rest_post_dispatch', 'wpapi_integration_force_http_content_type', 10, 3 );
```
Do you see the same error with the `_paging` assignment, or do you get different results?

**Comment by SotirisValogiannis** [comment]
@kadamwhite the plugin version is 1.4.9 and is the only active plugin. Have you set "Caching on" in the settings of the plugin?

Regarding your question as long as I have the plugin enabled with caching on the filter doesn't work. By the way did you mean `application/json` instead of `text/html` ? In any case it doesn't, `_paging` still `null`.

**Comment by kadamwhite** [comment]
@SotirisValogiannis Can you try installing a beta version of the plugin I just released:

```
npm install wpapi@beta
```

and, with the caching plugin enabled, let me know if the new version of the plugin corrects the issue?

Thanks!

**Comment by SotirisValogiannis** [comment]
@kadamwhite yes seems to works fine now!

**Comment by kadamwhite** [comment]
Released in v1.0.3! Thanks all.
